// allure-js-commons/dist/cjs/sync.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "ContentType", {
  enumerable: true,
  get: function get() {
    return _model.ContentType;
  }
});
Object.defineProperty(exports, "LabelName", {
  enumerable: true,
  get: function get() {
    return _model.LabelName;
  }
});
Object.defineProperty(exports, "LinkType", {
  enumerable: true,
  get: function get() {
    return _model.LinkType;
  }
});
Object.defineProperty(exports, "Severity", {
  enumerable: true,
  get: function get() {
    return _model.Severity;
  }
});
Object.defineProperty(exports, "Stage", {
  enumerable: true,
  get: function get() {
    return _model.Stage;
  }
});
Object.defineProperty(exports, "Status", {
  enumerable: true,
  get: function get() {
    return _model.Status;
  }
});
Object.defineProperty(exports, "StatusByPriority", {
  enumerable: true,
  get: function get() {
    return _model.StatusByPriority;
  }
});
Object.defineProperty(exports, "allureId", {
  enumerable: true,
  get: function get() {
    return _syncFacade.allureId;
  }
});
Object.defineProperty(exports, "attachTrace", {
  enumerable: true,
  get: function get() {
    return _syncFacade.attachTrace;
  }
});
Object.defineProperty(exports, "attachment", {
  enumerable: true,
  get: function get() {
    return _syncFacade.attachment;
  }
});
Object.defineProperty(exports, "attachmentPath", {
  enumerable: true,
  get: function get() {
    return _syncFacade.attachmentPath;
  }
});
Object.defineProperty(exports, "description", {
  enumerable: true,
  get: function get() {
    return _syncFacade.description;
  }
});
Object.defineProperty(exports, "descriptionHtml", {
  enumerable: true,
  get: function get() {
    return _syncFacade.descriptionHtml;
  }
});
Object.defineProperty(exports, "displayName", {
  enumerable: true,
  get: function get() {
    return _syncFacade.displayName;
  }
});
Object.defineProperty(exports, "epic", {
  enumerable: true,
  get: function get() {
    return _syncFacade.epic;
  }
});
Object.defineProperty(exports, "feature", {
  enumerable: true,
  get: function get() {
    return _syncFacade.feature;
  }
});
Object.defineProperty(exports, "globalAttachment", {
  enumerable: true,
  get: function get() {
    return _syncFacade.globalAttachment;
  }
});
Object.defineProperty(exports, "globalAttachmentPath", {
  enumerable: true,
  get: function get() {
    return _syncFacade.globalAttachmentPath;
  }
});
Object.defineProperty(exports, "globalError", {
  enumerable: true,
  get: function get() {
    return _syncFacade.globalError;
  }
});
Object.defineProperty(exports, "historyId", {
  enumerable: true,
  get: function get() {
    return _syncFacade.historyId;
  }
});
Object.defineProperty(exports, "issue", {
  enumerable: true,
  get: function get() {
    return _syncFacade.issue;
  }
});
Object.defineProperty(exports, "label", {
  enumerable: true,
  get: function get() {
    return _syncFacade.label;
  }
});
Object.defineProperty(exports, "labels", {
  enumerable: true,
  get: function get() {
    return _syncFacade.labels;
  }
});
Object.defineProperty(exports, "layer", {
  enumerable: true,
  get: function get() {
    return _syncFacade.layer;
  }
});
Object.defineProperty(exports, "link", {
  enumerable: true,
  get: function get() {
    return _syncFacade.link;
  }
});
Object.defineProperty(exports, "links", {
  enumerable: true,
  get: function get() {
    return _syncFacade.links;
  }
});
Object.defineProperty(exports, "logStep", {
  enumerable: true,
  get: function get() {
    return _syncFacade.logStep;
  }
});
Object.defineProperty(exports, "owner", {
  enumerable: true,
  get: function get() {
    return _syncFacade.owner;
  }
});
Object.defineProperty(exports, "parameter", {
  enumerable: true,
  get: function get() {
    return _syncFacade.parameter;
  }
});
Object.defineProperty(exports, "parentSuite", {
  enumerable: true,
  get: function get() {
    return _syncFacade.parentSuite;
  }
});
Object.defineProperty(exports, "severity", {
  enumerable: true,
  get: function get() {
    return _syncFacade.severity;
  }
});
Object.defineProperty(exports, "step", {
  enumerable: true,
  get: function get() {
    return _syncFacade.step;
  }
});
Object.defineProperty(exports, "story", {
  enumerable: true,
  get: function get() {
    return _syncFacade.story;
  }
});
Object.defineProperty(exports, "subSuite", {
  enumerable: true,
  get: function get() {
    return _syncFacade.subSuite;
  }
});
Object.defineProperty(exports, "suite", {
  enumerable: true,
  get: function get() {
    return _syncFacade.suite;
  }
});
Object.defineProperty(exports, "tag", {
  enumerable: true,
  get: function get() {
    return _syncFacade.tag;
  }
});
Object.defineProperty(exports, "tags", {
  enumerable: true,
  get: function get() {
    return _syncFacade.tags;
  }
});
Object.defineProperty(exports, "testCaseId", {
  enumerable: true,
  get: function get() {
    return _syncFacade.testCaseId;
  }
});
Object.defineProperty(exports, "tms", {
  enumerable: true,
  get: function get() {
    return _syncFacade.tms;
  }
});
var _syncFacade = require("./syncFacade.js");
var _model = require("./model.js");
