// allure-playwright/dist/esm/utils.js
function _typeof(o) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (o) { return typeof o; } : function (o) { return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o; }, _typeof(o); }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function _defineProperty(e, r, t) { return (r = _toPropertyKey(r)) in e ? Object.defineProperty(e, r, { value: t, enumerable: !0, configurable: !0, writable: !0 }) : e[r] = t, e; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == _typeof(i) ? i : i + ""; }
function _toPrimitive(t, r) { if ("object" != _typeof(t) || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != _typeof(i)) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); }
import { Stage, Status } from "allure-js-commons";
import { getMessageAndTraceFromError } from "allure-js-commons/sdk";
import { getWorstTestStepResult } from "allure-js-commons/sdk/reporter";
export var AFTER_HOOKS_ROOT_STEP_TITLE = "After Hooks";
export var BEFORE_HOOKS_ROOT_STEP_TITLE = "Before Hooks";
export var statusToAllureStats = function statusToAllureStats(status, expectedStatus) {
  if (status === "skipped") {
    return Status.SKIPPED;
  }
  if (status === "timedOut") {
    return Status.BROKEN;
  }
  if (status === expectedStatus) {
    return Status.PASSED;
  }
  return Status.FAILED;
};
export var isDescendantOfStepWithTitle = function isDescendantOfStepWithTitle(step, title) {
  var parent = step.parent;
  while (parent) {
    if (parent.title === title) {
      return true;
    }
    parent = parent.parent;
  }
  return false;
};
export var isAfterHookStep = function isAfterHookStep(step) {
  return isDescendantOfStepWithTitle(step, AFTER_HOOKS_ROOT_STEP_TITLE);
};
export var isBeforeHookStep = function isBeforeHookStep(step) {
  return isDescendantOfStepWithTitle(step, BEFORE_HOOKS_ROOT_STEP_TITLE);
};
export var diffEndRegexp = /-((expected)|(diff)|(actual))\.png$/;
export var normalizeHookTitle = function normalizeHookTitle(title) {
  return title.replace(/^[aA]ttach\s"(.+)"$/, "$1");
};
export var normalizeAttachStepTitle = function normalizeAttachStepTitle(title) {
  var match = title.match(/["'`]{1}(.+?)["'`]{1}/);
  if (match !== null && match !== void 0 && match[1]) {
    return match[1];
  }
  return title.replace(/^(test\.)?attach\s*/i, "").trim();
};
export var getSkipAnnotation = function getSkipAnnotation(step) {
  return step.annotations.find(function (annotation) {
    return annotation.type === "skip";
  });
};
export var getStepStatusData = function getStepStatusData(step, nestedStatus) {
  if (step.error) {
    return {
      status: Status.FAILED,
      statusDetails: _objectSpread({}, getMessageAndTraceFromError(step.error))
    };
  }
  var skipAnnotation = getSkipAnnotation(step);
  if (skipAnnotation) {
    return {
      status: Status.SKIPPED,
      statusDetails: skipAnnotation.description ? {
        message: skipAnnotation.description
      } : undefined
    };
  }
  return {
    status: nestedStatus !== null && nestedStatus !== void 0 ? nestedStatus : Status.PASSED
  };
};
export var finalizeStepResult = function finalizeStepResult(stepResult, step) {
  var _getWorstTestStepResu;
  var _ref = (_getWorstTestStepResu = getWorstTestStepResult(stepResult.steps)) !== null && _getWorstTestStepResu !== void 0 ? _getWorstTestStepResu : {},
    _ref$status = _ref.status,
    status = _ref$status === void 0 ? Status.PASSED : _ref$status;
  var _getStepStatusData = getStepStatusData(step, status),
    nextStatus = _getStepStatusData.status,
    statusDetails = _getStepStatusData.statusDetails;
  stepResult.status = nextStatus;
  stepResult.stage = Stage.FINISHED;
  if (statusDetails) {
    // Preserve any previously collected fields and only overwrite values provided by the current step outcome.
    stepResult.statusDetails = _objectSpread(_objectSpread({}, stepResult.statusDetails), statusDetails);
  }
};
